### DISCUSSION TITLE
AI improvements master topic

### DISCUSSION THREAD

**Discussion by xezon** [TheSuperHackers/GeneralsGamePatch#685] [General]
* Investigate what can be done to fix AI bugs and mismatches
* Explore what kind of new AI features and AI difficulties could be added
* Decide whether or not any of the original AI difficulties (Easy, Medium, Hard) should be tweaked or replaced

Potentially include high level AI decision in community survey to get a better overview of what players want from AI.

[Advanced AI Mod changes & review](https://github.com/xezon/GeneralsGamePatch/blob/main/Patch104pZH/Design/Tasks/advancedai_tasks.txt)

[AI survey results](https://github.com/xezon/GeneralsGamePatch/tree/main/Patch104pZH/Design/Survey/Results/2_ai)

**Comment by xezon** [comment]
Zeke:
> Mismatch: One of them is projectile weapons lacking min range
> If a unit fires a projectile weapon to close to itself it causes a mismatch
> min range is more of a bandaid than an actual solution, but there doesn't seem to be any better methods to fix it, aside from players consciously keeping proper distance

#212

**Comment by ZekeDlyoung** [comment]
Fix AI bugs and mismatches I agree with as this will drastically improve performance and make crashes less frequent

We should probably hold off on increasing the actual AI difficulty, the AI will naturally get harder when most of the bugs are fixed

At most I agree with Easy AI getting tweaked to send more units, but the rest should stay relatively the same. If people want more difficulty they can fight other people, or fight against more AIs

I would also be wary of making a poll about AI difficulty, because we've had such polls in the past, and what usually ends up happening is that the PvP crowd votes for harder AI because the current AI bores them, but all this ends up doing is make the game/mod less appealing to comp stompers who are fine with the current AI

**Comment by commy2** [comment]
Apparently the minrange mismatch fix would involve giving tank weapons a minrange of 5. This does affect PVP games as well.

In my opinion, if a fix for compstomp affects pvp in a way that is not desirable, then we should not apply it.

**Comment by xezon** [comment]
I am skeptical about not involving community to make decision regarding AI difficulty. We can design survey to get data on who is a regular AI Stomper and who is not. This way we can weigh the results accordingly.

There are some considerations to make:
* Is a harder AI option appealing to pull more casual players to play more Multiplayer?
* Is a harder AI option appealing to pull more PvP players into the AI genre (good for retention)?
* Would harder AI option help the popularity of this Patch overall?

I understand current AI players are happy the way it is (except for the mismatches), but this is just logical, because if they were not happy they would not be playing it right? We don't know what happens if AI becomes better. You may suddenly see twice as many AI Stompers if it gives them a better experience.

Jundiyy mentioned Easy AI is useless. Personally I am no AI regular, but I do agree that Easy AI does not do all that much and I wonder if it can be repurposed in the following way:
* MediumAI becomes EasyAI
* HardAI becomes MediumAI
* BrutalAI (new) becomes HardAI

If players still wanted to stomp EasyAI, they could always play Original version. Our Patch install will always grant full access to Original version natively.

**Comment by ZekeDlyoung** [comment]
"We can design survey to get data on who is a regular AI Stomper and who is not. This way we can weigh the results accordingly."

Yeah this is what I meant, I didn't say don't make a poll, I just said be wary/carefull, as other players love trolling comp stompers.

And yes easy AI currently is useless

**Comment by xezon** [comment]
Jundiyy:
> **Mismatch:**
> Also mismatch issues with multiplayer AI:
> Bulling scaffolds in the AI base within the first minute or so
> Attacking/Rushing the AI within the first 2 minutes or so
> **Hard AI:**
> 1 fix for the Hard AI could be them not slowing down with attacks? At the minimum they keep up the attacks even if they aren't improved otherwise

**Comment by MTKing4** [comment]
> Jundiyy mentioned Easy AI is useless. Personally I am no AI regular, but I do agree that Easy AI does not do all that much and I wonder if it can be repurposed in the following way:

> + MediumAI becomes EasyAI

> + HardAI becomes MediumAI

> + BrutalAI (new) becomes HardAI

Agreed

**Comment by ThePredatorBG** [comment]
I recommend making a poll about AI changes, so we get as many people's feedback as possible. But maybe in the future combined with other non-AI questions, because we recently had our first poll.

I believe [Advanced AI mod 0.98 improved by AEI v2.8](https://www.moddb.com/mods/advanced-ai-mod-for-zero-hour-improved-by-aei/downloads/advanced-ai-mod-098-improved-by-aei-v28) is not only a good source of reference, but also a good candidate for becoming the starting point of our AI. There is a Readme.txt with all the changes made in the archive. The mod changes some of the .ini files to help some of the scripts work, like the Bomb Truck Disguise ability, but those .ini changes can be easily adapted. The mod not only increases the difficulty, but also fixes many issues with the original scripts. I suggest that we ask AEI for permissions to adapt his AI. Tremendous work has been done. Additional difficulty changes, if required, can be made more easily than going the same path over again for the sake of keeping close track of difficulty, ensuring it doesn't go up.

I don't have any observations about AI-triggered mismatches, but it always seemed like AI does increase mismatch occurrence chances. Jundiyy's statements about early mismatch are interesting and can be easily tested. It would be great if we discover which scripts or teams are buggy and cause mismatch.

**Comment by Kesanov** [comment]
> I understand current AI players are happy the way it is (except for the mismatches), but this is just logical, because if they were not happy they would not be playing it right? We don't know what happens if AI becomes better. You may suddenly see twice as many AI Stompers if it gives them a better experience.

I am of the same opinion. For me, there is no reason to play with AI, when it's not challenging enough.

**Comment by ZekeDlyoung** [comment]
> For me, there is no reason to play with AI, when it's not challenging enough.

That may be true for some people, but for others RTS is about the sandbox experience, using the time to unwind from a hard day's work not be more stressed. People can find challenges by fighting other people, but there is no alternative other than AI battles for recreational gaming. Yes, people can just switch back to normal ZH if they just want to comp stomp and chill, but as mentioned the normal ZH AI has many bugs. I for one would just like to see first just the normal ZH AI with fixes, that could already be challenging enough

**Comment by MTKing4** [comment]
We could make the advanced AI a tickable option in the launcher tool then, you can launch our patch with it on or off but keep it on by default, similar to how contra has options for many things like switching different hotkey sets

**Comment by xezon** [comment]
> We could make the advanced AI a tickable option in the launcher tool then, you can launch our patch with it on or off but keep it on by default, similar to how contra has options for many things like switching different hotkey sets

I am very skeptical of this, because it would increase the complexity to arrange a non-mismatching match session. Any change that alters game state cannot be a simple user choice. Any change that alters local settings/setup only can be a simple user choice.

**Comment by MTKing4** [comment]
Then make the normal AI option disabled by default for online games

The option could be called something like this:

✅ Old Hard AI (Single player)

**Comment by xezon** [comment]
A survey will give more clarity on player preferences.

**Comment by xezon** [comment]
Enlima29:
> I wonder if there should be these three difficulties based on player intentions:
> - Dummy AI (for target practice and testing purposes)
> - Human-like AI (for regular PvE enemies)
> - Insanely Cheating AI (for those who like taking challenges against unfair enemies)
>
> **also iirc AI players have hidden handicap/modifier values defined in the gamedata.ini, those should be taken into consideration when designing the AI**

**Comment by xezon** [comment]
Enlima29:
> the only time AI difficulties matter in multiplayer is when you decide to have coop games against or with AI
fixing (and improving) the AI will only majorly affecting those who usually singleplay
though regardless it is in multiplayer or solo play, they play against AI with different motives in mind
that's why I propose rather than having traditional escalating AI difficulties (Easy-Medium-Hard-Brutal) it better they were repurposed as AI personalities
> * Dummy AI will be largely passive and turtle as their purpose is to be target practice for testing or as introducing enemy for rather new players
> * Normal AI will be your primary choice when you are playing, as they can simulate what players usually do in multiplayer, this also can serveas training AI to prepare you to jump in online games against human players
> * Lastly, Insane AI will be designed for compstomper that like to seek challenges against super hard AI. this one is allowed to cheat and all which gives them unfair advantages
>
> I think this way it can please both new and experienced players, whatever their intention playing against the AI is
if you want to encite players to play online games more, I believe the best way is to make sure players in the match having similar skill levels so nobody will get discouraged when they suddenly matched against far more experienced players and lose within few minutes, thus proper matchmaking is needed to have fairer matches for everyone, but I think that is impossible to organize within the current form of ZH game itself.
though, even with matchmaking system it would be useless if there is no steady new player pool
> ZH is an old game that idk if people still care about or not, without playerbase regeneration this game multiplayer will surely die in time where either current playerbase quitting, or just filled with experienced players making it hard for new players to climb up to compete against the current playerbase skill.

**Comment by 3CG-Generals** [comment]
I just want to say, that you will not be able to script an AI which is acting comparible to a human player without little cheats for AI. Example, if you change build order for base building (even AiData.ini for positions), it will always choose next free dozer and not the nearist free dozer in relation to the position of the next structure which shall be build up next. In other words, dozers drive to other side of base and build there and looses time. And there are more examples of it. What can be improved is for example use of laserlock and switch to certain kind of units during attacks... Which will already make an impact on difficulty.

**Comment by xezon** [comment]
Jundiyy:
> I think one thing to keep in mind is to keep AI base build sizes the same, or perhaps a bit smaller if possible. Otherwise the AI would not work as well anymore on older maps if they do not have a sufficient base size.

**Comment by xezon** [comment]
Follow up on [AI survey results](https://github.com/xezon/GeneralsGamePatch/tree/main/Patch104pZH/Design/Survey/Results/2_ai).

* Players tend to think Easy AI is boring
* Players tend to think Medium AI is boring or ok, but not exciting
* Players tend to think Hard AI is more exciting than boring
* Many players see issues with AI that they would like to see improved
* The vast majority of players would potentially play more vs AI if it was more challenging
* 75%+ players are ok with both replacing Easy AI and/or Hard AI for more challenging variant

We need to figure out how we want to approach tackling AI improvements.

Which AI files do we use as a base?
* Original Game AI Scripts?
* Advanced AI Mod Scripts?
* None?
* Other?

What high level actions do we need to focus early on?
* Eliminate bugs & mismatches?
* Eliminate poor performing scripts, such as "Hunt"?
* Improve AI base builds?
* Improve AI unit composition?
* Improve AI attack and defend logic?
* Adjust Official Maps in a way that allows to use more clever AI pathing?
* Reduce AI cheating?
* How can AI be made smarter as to where it just does not waste all its forces into a wall of Laser Turrets?

**Reply by DoMiNaToRuk2** [reply]
Please make the AI as strong as possible. Please make the buildings actually build from start, often it freezes making a supply. 

Maybe add brutal option and copy the data from
Advanced ai mod. 

This should be a definite

**Comment by samosaphile** [comment]


**Reply by penfriendz** [reply]
https://github.com/TheSuperHackers/GeneralsGamePatch/issues/657

**Reply by samosaphile** [reply]
Ah, completely missed that but did try searching first. Deleted since duplicate post.
